### PR TITLE
Print standards report

### PR DESCRIPTION
**Pull request by dmcavoy**
This PR got away from me a little bit so its long. 

Creates the print view for the standards report. This report is generated when you use the "Generate PDF report" button. The report is generated on a separate page that is at `/teacher_dashbaord/sections/#/standards_report`. From that page you can click print report and that will give you the option to print it or save it based on your browser printing settings.

Some future work:
- Because we are opening a new page some of the redux state information is not getting transferred so I need to work through how to make sure the print page has all the correct information. (This includes the teacher comment which is being saved to redux but not sent to the new page.)
- When you actually print the report there is some CSS work to do there.
- Lessons completed and lessons available are just hard coded right now.
- Writing tests for all the new features
- Figure out how to make the standards report load faster

standards-report

## Links

- [spec](https://docs.google.com/document/d/1xlm_syljGXIapl72GaKxN63V7kt6syHyvWHuqZhTCfE/edit#heading=h.jnuuhrkafjqp)
- [jira](https://codedotorg.atlassian.net/browse/LP-1126)

## Testing story

I need to follow up with this work as well.

# Reviewer Checklist:

- [ ] Tests provide adequate coverage
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: apps/i18n/common/en_us.json [hunk 1]
```diff
@@ -806,6 +811,7 @@
   "justDidHourOfCode": "I just did the #HourOfCode - check it out! @codeorg",
   "justDidHourOfCodeDonor": "I just did the #HourOfCode - check it out! (Thanks {donor_twitter} for supporting @codeorg)",
   "keepPlaying": "Keep playing",
+  "key": "Key: ",
```
Comment (clareconstantine) [inline apps/i18n/common/en_us.json]: I'm not sure how strict this pattern is, but do you want this to be called `keyWithColon`?

File: apps/i18n/common/en_us.json [hunk 2]
```diff
@@ -1335,6 +1347,7 @@
   "schoolNotFoundDescription": "Sorry, we couldn't find your school. Please enter information below.",
   "score": "score",
   "section": "Section",
+  "sectionAndName": "Section:",
```
Comment (clareconstantine) [inline apps/i18n/common/en_us.json]: same q here - `sectionWithColon` seems more accurate

File: apps/i18n/common/en_us.json [hunk 3]
```diff
@@ -1440,9 +1453,14 @@
   "standardsAndFrameworkDescription": "Information about curriculum standards alignment for Code.org courses.",
   "standardsGetInvolved": "Get others involved in CS education",
   "standardsGetInvolvedDetails": "Learn how to get your [administrators]({adminLink}), [parents]({parentLink}), and [fellow teachers]({teacherLink}) involved in CS education.",
+  "standardsGetInvolvedDetailsForPrint": "Learn more about how you can support computer science education as an [administrators]({adminLink}), [parents]({parentLink}), or [teachers]({teacherLink}).",
```
Comment (clareconstantine) [inline apps/i18n/common/en_us.json]: nit: looks like 'administrators, parents, or teachers,' shouldn't be plural

File: apps/i18n/common/en_us.json [hunk 4]
```diff
@@ -1573,6 +1594,7 @@
   "useToViewList3": "**Generate a PDF report** to save or share with administrators or parents",
   "usingHints": "Using hints",
   "usingTooManyBlocks": "Using too many blocks",
+  "uncompeltedLessons": "Lesson not completed",
```
Comment (clareconstantine) [inline apps/i18n/common/en_us.json]: nit: typo - "uncompleted"

File: apps/src/templates/sectionProgress/standards/StandardsReport.jsx [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+    printWindow.addEventListener('load', event => {
+      printWindow.print();
+    });
+
+    printWindow.document.write(
+      `<html><head><title>${i18n.printReportWindowTitle({
+        sectionName: this.props.section.name
+      })}</title></head>`
+    );
+    printWindow.document.write('<body>'); //onafterprint="self.close()"
```
Comment (clareconstantine) [inline apps/src/templates/sectionProgress/standards/StandardsReport.jsx]: is this comment still needed?

File: apps/src/templates/sectionProgress/standards/sectionStandardsProgressRedux.js [hunk 6]
```diff
@@ -55,6 +68,14 @@ export function getLessonsCompletedByStandardForScript(script) {
   return lessonCompletedByStandard;
 }
 
+export function getNumberLessonsCompleted(script) {
+  return 5;
```
Comment (clareconstantine) [inline apps/src/templates/sectionProgress/standards/sectionStandardsProgressRedux.js:72]: where do these numbers come from? should these be dynamically calculated? if not, could they be stored as constants and maybe have a comment explaining why it's always 5 (or 10, for lessons in course)?

**Review by clareconstantine — APPROVED** [review]
A few very minor comments but overall this looks good to me! Nice work on this task!

File: apps/src/templates/sectionProgress/standards/sectionStandardsProgressRedux.js (hunk 6 above)
Comment (dmcavoy) [reply apps/src/templates/sectionProgress/standards/sectionStandardsProgressRedux.js:72]: This is just fake data right now. I was waiting for Erin to finish with the lesson completion logic before actually trying to calculate this.

File: apps/src/redux/scriptSelectionRedux.js [hunk 7]
```diff
@@ -50,6 +50,18 @@ export const getSelectedScriptFriendlyName = state => {
...
+export const getSelectedScriptDescription = state => {
+  const scriptId = state.scriptSelection.scriptId;
+  if (!scriptId) {
+    return null;
+  }
+
+  const scripts = state.scriptSelection.validScripts;
+  const script = scripts.find(script => script.id === scriptId);
+  return script ? script.description : null;
+};
```
Comment (Erin007) [inline apps/src/redux/scriptSelectionRedux.js:63]: Where is the description property coming from? I'd expect to see "description" here: https://github.com/code-dot-org/code-dot-org/blob/2364fe89fee1212406f491d41819af6d9869a5a8/apps/src/redux/scriptSelectionRedux.js#L56 if it were available on the validScript object.

File: apps/src/templates/sectionProgress/standards/StandardsLegendForPrint.jsx [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+  },
+  keyItems: {
+    display: 'flex',
+    flexDirection: 'row',
+    alignItems: 'center',
+    margin: '0 10px'
+  }
+};
+
+export default class StandardsLegend extends Component {
```
Comment (Erin007) [inline apps/src/templates/sectionProgress/standards/StandardsLegendForPrint.jsx]: `StandardsLegend` -> `StandardsLegendForPrint` ?

File: apps/src/templates/sectionProgress/standards/StandardsLegendForPrint.jsx [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+        <span style={styles.keyItems}>
+          {i18n.completedLessons()}
+          <ProgressBoxForLessonNumber completed={true} lessonNumber={1} />
+        </span>
+        <span style={styles.keyItems}>
+          {i18n.uncompletedLessons()}
+          <ProgressBoxForLessonNumber completed={false} lessonNumber={1} />
+        </span>
+      </div>
+    );
```
Comment (Erin007) [inline apps/src/templates/sectionProgress/standards/StandardsLegendForPrint.jsx:33]: Probably more of a product question, but would it be helpful to make the standards view and the printed report more similar so we can re-use more components/strings?

File: apps/src/templates/sectionProgress/standards/StandardsReport.jsx [hunk 10]
```diff
@@ -0,0 +1,206 @@
...
+
+  getLinkToOverview() {
+    const {scriptData, section} = this.props;
+    return scriptData ? `${scriptData.path}?section_id=${section.id}` : null;
+  }
+
+  printReport = () => {
+    const printArea = document.getElementById('printArea').outerHTML;
+    // Adding a unique ID to the window name allows for multiple instances of this window
+    // to be open at once without affecting each other.
```
Comment (Erin007) [inline apps/src/templates/sectionProgress/standards/StandardsReport.jsx:87]: clever!

File: apps/src/templates/sectionProgress/standards/StandardsReport.jsx [hunk 11]
```diff
@@ -0,0 +1,206 @@
...
+            </div>
+          </div>
+        </div>
+        <Button
+          onClick={this.printReport}
+          color={Button.ButtonColor.orange}
+          text={i18n.printReport()}
+          size={'narrow'}
+          style={styles.button}
+        />
```
Comment (Erin007) [inline apps/src/templates/sectionProgress/standards/StandardsReport.jsx]: This button at the top and bottom of the page? Worth extracting if we're going to render twice?

File: apps/src/templates/sectionProgress/standards/StandardsReportCurrentCourseInfo.jsx [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+            <a href={linkToOverview} style={styles.scriptLink}>
+              {this.props.scriptFriendlyName}
+            </a>
+          </h3>
+          <p>{this.props.scriptDescription} </p>
+          <ul>
+            <li>
+              <SafeMarkdown
+                markdown={i18n.mapsToCSTAStandards({
+                  cstaLink: 'https://www.csteachers.org/page/standards'
```
Comment (Erin007) [inline apps/src/templates/sectionProgress/standards/StandardsReportCurrentCourseInfo.jsx]: I wonder if this is worth putting in a constant to share across uses?

**Review by Erin007 — APPROVED** [review]
Holy Moly! Awesome start! I'm looking forward to the follow-ups you outlined in the description, but this is really coming along.  I have a broad question about whether we can share more between the report and the standards view, but not blocking and maybe dependent on a chat with Amanda.

File: apps/src/redux/scriptSelectionRedux.js (hunk 7 above)
Comment (dmcavoy) [reply apps/src/redux/scriptSelectionRedux.js:63]: I got the description in script_constants.rb but I don't know that thats the right place for it and now its breaking some tests so I'm gonna take another look.

File: apps/src/templates/sectionProgress/standards/sectionStandardsProgressRedux.js (hunk 6 above)
Comment (clareconstantine) [reply apps/src/templates/sectionProgress/standards/sectionStandardsProgressRedux.js:72]: that makes sense! thanks!

**Pull request merged**